theano/sandbox/cuda/cula.py

```diff
@@ -140,3 +143,142 @@ def cula_gpu_solve(A_, b_, trans='T'):
         return thunk
 
 gpu_solve = GpuSolve()
+
+
+class GpuCholesky(GpuOp):
+    """
+    CULA GPU Choleksy factorisation op.
+
+    Given a real positive definite matrix `A` returns either a lower
+    triangular matrix `L` such that `A == dot(L, L.T)` if `lower == True`
+    else returns an upper triangular matrix `U` such that `A == dot(U.T, U)`
+    if `lower == False`.
+
+    Parameters
+    ----------
+    lower
+        Whether to return a lower rather than upper triangular decomposition.
+
+    """
+
+    __props__ = ('lower',)
+
+    def __init__(self, lower=True):
+        self.lower = lower
+        super(GpuCholesky, self).__init__()
+
+    def output_type(self, inp):
+        return CudaNdarrayType(broadcastable=[False] * inp.type.ndim)
+
+    def make_node(self, inp):
+        inp = as_cuda_ndarray_variable(inp)
+
+        assert inp.ndim == 2
+        return theano.Apply(self, [inp], [self.output_type(inp)()])
+
+    def make_thunk(self,
+                   node,
+                   storage_map, _,
+                   no_recycling=[]):
+
+        # Initialize CULA the first time it is needed
+        global cula_initialized
+
+        if not cula_available:
+            raise RuntimeError('Cula is not available and '
+                               'GpuCholesky Op can not be constructed.')
+
+        if not pycuda_available:
+            raise RuntimeError('pycuda is not available and '
+                               'GpuCholesky Op can not be constructed.')
+
+        if not cula_initialized:
+            cula.culaInitialize()
+            cula_initialized = True
+
+        # import util function here to avoid circular import errors
+        from theano.misc.pycuda_utils import to_gpuarray
+
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        def thunk():
+
+            # Matrix to decompose is first (and only) input
+            A = inputs[0][0]
+
+            # Cholesky decomposition of matrix is first (and only) output
+            A_chol = outputs[0]
+
+            # CULA Cholesky function is destructive as it assigns calculated
+            # decomposition to lower / upper triangle of input array therefore
+            # force copy of array with allocation of a new C-contiguous buffer.
+            # CULA operation assumes Fortran ordering however input matrix must
+            # be symmetric therefore no need to alter order.
```

Comment:
The Choleksy decomposition is only defined for positive definite matrices, which for real matrices are a subset of the symmetric matrices. If a non-symmetric and so non-positive-definite input is passed the underlying CULA function will raise an exception as the Cholesky decomposition algorithm will fail. Given an exception will be raised anyway without any additional checking overhead for non-symmetric inputs it didn't seem worth adding any explicit checks for symmetry. 

I guess the comment is not strictly correct however so I'll edit to make it more clear - perhaps it would be better to say any valid input will be symmetric? It is possible a non-symmetric matrix input will be passed to the op however it does not matter whether it uses C or Fortan ordering here as in either case the matrix will remain non-symmetric and end up raising an exception when it is attempted to calculate the decomposition. 

I'll also add an unit test to ensure non-symmetric inputs cause an exception to be raised in addition to the current one for a non-positive-definite but symmetric input.